// @thearchitech.xyz/genome-transformer/dist/transformers/framework-stripper.js
import { BaseGenomeTransformer } from '../core/genome-transformer.js';
/**
 * Removes framework adapter modules from the module list.
 * Frameworks are handled via project.apps and should not execute as modules.
 */
export class FrameworkStripper extends BaseGenomeTransformer {
    name = 'FrameworkStripper';
    description = 'Removes adapters/framework/* modules (frameworks are executed outside the module pipeline).';
    priority = 60; // run before ExecutionOrder/FinalValidator (which were ~70-90)
    canTransform(genome) {
        return Array.isArray(genome.modules) && genome.modules.some(m => m.id.startsWith('adapters/framework/'));
    }
    async transform(genome, context) {
        const before = genome.modules.length;
        const removedFrameworks = genome.modules.filter(m => m.id.startsWith('adapters/framework/')).map(m => ({
            id: m.id,
            frameworkId: m.id.replace('adapters/framework/', 'framework/'),
            parameters: m.parameters || {},
            features: m.features || {}
        }));
        const filtered = genome.modules.filter(m => !m.id.startsWith('adapters/framework/'));
        const removed = before - filtered.length;
        if (removed > 0) {
            context.logger.info('🧹 Removed framework modules from execution list', { removed, before, after: filtered.length });
        }
        if (removedFrameworks.length === 0) {
            return { ...genome, modules: filtered };
        }
        const existingFrameworks = ((genome.metadata || {}).bootstrap || {}).frameworks || [];
        const metadata = {
            ...genome.metadata,
            bootstrap: {
                ...((genome.metadata || {}).bootstrap || {}),
                frameworks: [...existingFrameworks, ...removedFrameworks]
            }
        };
        return { ...genome, modules: filtered, metadata };
    }
}
