// @thearchitech.xyz/genome-transformer/dist/transformers/parameter-distribution-transformer.js
/**
 * Parameter Distribution Transformer
 *
 * Transforms capability parameters into module-specific parameters
 * with validation to ensure modules can handle the distributed parameters.
 */
import { BaseGenomeTransformer } from '../core/genome-transformer.js';
import { ParameterDistributionService } from '../core/parameter-distribution.js';
export class ParameterDistributionTransformer extends BaseGenomeTransformer {
    name = 'Parameter Distribution';
    priority = 3; // After capability normalization and module expansion
    description = 'Distributes capability parameters to modules with validation';
    distributionService;
    constructor() {
        super();
        this.distributionService = new ParameterDistributionService({
            info: (message, meta) => console.log(`[INFO] ${message}`, meta || ''),
            debug: (message, meta) => console.log(`[DEBUG] ${message}`, meta || ''),
            warn: (message, meta) => console.log(`[WARN] ${message}`, meta || ''),
            error: (message, meta) => console.log(`[ERROR] ${message}`, meta || '')
        });
    }
    canTransform(genome) {
        // Only transform if genome has capabilities with parameters in the new structure
        if (!genome.capabilities)
            return false;
        return Object.values(genome.capabilities).some(config => {
            if (!config)
                return false;
            const cfg = config;
            // Check for new structure: adapter, frontend, techStack, shared, backend, database
            return !!((cfg.adapter && Object.keys(cfg.adapter).length > 0) ||
                (cfg.frontend && Object.keys(cfg.frontend).length > 0) ||
                (cfg.techStack && Object.keys(cfg.techStack).length > 0) ||
                (cfg.shared && Object.keys(cfg.shared).length > 0) ||
                (cfg.backend && Object.keys(cfg.backend).length > 0) ||
                (cfg.database && Object.keys(cfg.database).length > 0) ||
                // Legacy support
                (cfg.parameters && Object.keys(cfg.parameters).length > 0));
        });
    }
    async transform(genome, context) {
        this.logStep(context, 'Distributing capability parameters to modules', {
            capabilities: Object.keys(genome.capabilities || {}).length,
            modules: genome.modules.length
        });
        try {
            const result = await this.distributionService.distributeParameters(genome, context.marketplacePath);
            if (!result.success) {
                context.logger.error('Parameter distribution failed', {
                    errors: result.errors,
                    warnings: result.warnings
                });
                // Log errors but don't fail the transformation
                // Let the final validator handle parameter validation
                context.logger.warn('Parameter distribution had errors, but continuing with transformation', {
                    errors: result.errors.length,
                    warnings: result.warnings.length
                });
            }
            this.logSuccess(context, 'Parameter distribution complete', {
                modules: result.modules.length,
                errors: result.errors.length,
                warnings: result.warnings.length
            });
            return {
                ...genome,
                modules: result.modules
            };
        }
        catch (error) {
            context.logger.error('Parameter distribution failed', {
                error: error instanceof Error ? error.message : 'Unknown error'
            });
            // Return genome unchanged on error
            return genome;
        }
    }
    async validate(genome) {
        const errors = [];
        const warnings = [];
        // Validate that capability parameters are properly structured
        if (genome.capabilities) {
            for (const [capabilityName, capabilityConfig] of Object.entries(genome.capabilities)) {
                if (!capabilityConfig)
                    continue;
                // Check if parameters object exists and is valid
                if (capabilityConfig.parameters && typeof capabilityConfig.parameters !== 'object') {
                    errors.push(`Capability ${capabilityName} has invalid parameters: must be an object`);
                }
                // Check for common parameter validation
                if (capabilityConfig.parameters) {
                    for (const [paramName, paramValue] of Object.entries(capabilityConfig.parameters)) {
                        if (paramValue === null || paramValue === undefined) {
                            warnings.push(`Capability ${capabilityName} has null/undefined parameter: ${paramName}`);
                        }
                    }
                }
            }
        }
        return {
            valid: errors.length === 0,
            errors,
            warnings
        };
    }
}
