// @tamagui/config/dist/esm/v5.mjs
import { shorthands } from "@tamagui/shorthands/v5";
import { tokens, defaultThemes } from "@tamagui/themes/v5";
import { animations } from "./v5-animations.mjs";
import { fonts } from "./v5-fonts.mjs";
import { media, mediaQueryDefaultActive } from "./v5-media.mjs";
import { shorthands as shorthands2 } from "@tamagui/shorthands/v5";
import { createThemes } from "@tamagui/theme-builder";
import { tamaguiThemes, tokens as tokens2 } from "@tamagui/themes/v5";
import { animations as animations2 } from "./v5-animations.mjs";
import { createSystemFont, fonts as fonts2 } from "./v5-fonts.mjs";
import { breakpoints, media as media2, mediaQueryDefaultActive as mediaQueryDefaultActive2 } from "./v5-media.mjs";
import { defaultThemes as defaultThemes2 } from "@tamagui/themes/v5";
const selectionStyles = theme => theme.color5 ? {
    backgroundColor: theme.color5,
    color: theme.color11
  } : null,
  settings = {
    mediaQueryDefaultActive,
    defaultFont: "body",
    fastSchemeChange: !0,
    shouldAddPrefersColorThemes: !0,
    allowedStyleValues: "somewhat-strict-web",
    addThemeClassName: "html",
    // allow two inverses (tooltips, etc)
    // TODO on inverse theme changes
    maxDarkLightNesting: 2,
    styleCompat: "react-native"
  },
  defaultConfig = {
    animations,
    media,
    shorthands,
    themes: defaultThemes,
    tokens,
    fonts,
    selectionStyles,
    settings
  };
export { animations2 as animations, breakpoints, createSystemFont, createThemes, defaultConfig, fonts2 as fonts, media2 as media, mediaQueryDefaultActive2 as mediaQueryDefaultActive, selectionStyles, settings, shorthands2 as shorthands, tamaguiThemes, defaultThemes2 as themes, tokens2 as tokens };
