// @tamagui/config/dist/cjs/v5.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);
var v5_exports = {};
__export(v5_exports, {
  animations: () => import_v5_animations2.animations,
  breakpoints: () => import_v5_media2.breakpoints,
  createSystemFont: () => import_v5_fonts2.createSystemFont,
  createThemes: () => import_theme_builder.createThemes,
  defaultConfig: () => defaultConfig,
  fonts: () => import_v5_fonts2.fonts,
  media: () => import_v5_media2.media,
  mediaQueryDefaultActive: () => import_v5_media2.mediaQueryDefaultActive,
  selectionStyles: () => selectionStyles,
  settings: () => settings,
  shorthands: () => import_v53.shorthands,
  tamaguiThemes: () => import_v54.tamaguiThemes,
  themes: () => import_v55.defaultThemes,
  tokens: () => import_v54.tokens
});
module.exports = __toCommonJS(v5_exports);
var import_v5 = require("@tamagui/shorthands/v5"), import_v52 = require("@tamagui/themes/v5"), import_v5_animations = require("./v5-animations"), import_v5_fonts = require("./v5-fonts"), import_v5_media = require("./v5-media"), import_v53 = require("@tamagui/shorthands/v5"), import_theme_builder = require("@tamagui/theme-builder"), import_v54 = require("@tamagui/themes/v5"), import_v5_animations2 = require("./v5-animations"), import_v5_fonts2 = require("./v5-fonts"), import_v5_media2 = require("./v5-media"), import_v55 = require("@tamagui/themes/v5");
const selectionStyles = (theme) => theme.color5 ? {
  backgroundColor: theme.color5,
  color: theme.color11
} : null, settings = {
  mediaQueryDefaultActive: import_v5_media.mediaQueryDefaultActive,
  defaultFont: "body",
  fastSchemeChange: !0,
  shouldAddPrefersColorThemes: !0,
  allowedStyleValues: "somewhat-strict-web",
  addThemeClassName: "html",
  // allow two inverses (tooltips, etc)
  // TODO on inverse theme changes
  maxDarkLightNesting: 2,
  styleCompat: "react-native"
}, defaultConfig = {
  animations: import_v5_animations.animations,
  media: import_v5_media.media,
  shorthands: import_v5.shorthands,
  themes: import_v52.defaultThemes,
  tokens: import_v52.tokens,
  fonts: import_v5_fonts.fonts,
  selectionStyles,
  settings
};
